Support `wasm32-unknown-unknown` target

Spxg opened the pull request:
I don't know if it can be implemented this way.

gwenn commented:
I know nothing about wasm so we need to find someone kind enough to validate this PR.
And would you mind adding a dedicated job, something like https://github.com/rusqlite/rusqlite/pull/1010/files#diff-7829468e86c1cc5d5133195b5cb48e1ff6c75e3e9203777f6b2e379d9e4882b3R86
in your PR
Thanks.

gwenn commented:
Thanks for your time and explanations.

gwenn commented:
> Fixed CI for `clippy` and `minimal versions`. The reduced coverage is probably because I added two wasm-only tests and some comments.
> 
> By the way, why does the CI process require approval? It seems that it is not convenient to find problems and fix CI.

Maybe because a patched / malicious `build.rs`, you can do anything !

trevyn commented:
> By the way, why does the CI process require approval? It seems that it is not convenient to find problems and fix CI.

You can enable Actions on your fork and get CI to run inside your account instead of from the PR.

gwenn commented:
There is something wrong here:
https://github.com/rusqlite/rusqlite/actions
```
[-unknown-unknown target](https://github.com/rusqlite/rusqlite/actions/runs/13281257633)
CI #3544: Pull request https://github.com/rusqlite/rusqlite/pull/1643 synchronize by [Spxg](https://github.com/Spxg)
[Spxg:wasm-demo](https://github.com/Spxg/rusqlite/tree/refs/heads/wasm-demo)	
 8 hours ago
 In progress
```

gwenn commented:
> > we need to find someone kind enough to validate this PR.
> 
> Any suggestions?

Maybe @thomcc or @trevyn ?

lucyawrey commented:
This PR is exactly what I need for one of my projects, so I hope someone can validate it soon.

I can at least confirm @Spxg 's fork works without issue when used as a dependency in a project targeting `wasm32-unknown-unknown`. I only seem to be able to get it to work in-memory for `wasm` builds though, and am not sure how to get it to work with a persistent VFS.

lucyawrey commented:
@Spxg Thanks for the information!

Spxg commented:
> > > > we need to find someone kind enough to validate this PR.
> > >
> > >
> > > Any suggestions?
> >
> > Maybe @thomcc or @trevyn ?
>>
> It seems like they all don't have time to review it, so I thought of some questions you might be concerned about, I hope this can answer your questions.

@thomcc @gwenn @trevyn any other questions or suggestions?

burrbull commented:
I've compiled `rusqlite` from this branch. But I'm falling trying to open existing database.
I'm using `trunk serve` with `copy-dir`.
When I open connection with `OpenFlags::SQLITE_OPEN_READ_ONLY` I get `CannotOpen` error.
If I open connection with `OpenFlags::default()` it works but then it fails to open table which means that this is new database created somewhere else (I could not find where, maybe in-memory?).
Thanks.

Spxg commented:
> I've compiled `rusqlite` from this branch. But I'm falling trying to open existing database. I'm using `trunk serve` with `copy-dir`. When I open connection with `OpenFlags::SQLITE_OPEN_READ_ONLY` I get `CannotOpen` error. If I open connection with `OpenFlags::default()` it works but then it fails to open table which means that this is new database created somewhere else (I could not find where, maybe in-memory?). Thanks.

There is no native file system on the web, so VFS needs to be implemented. https://github.com/Spxg/sqlite-wasm-rs/blob/master/sqlite-wasm-rs/src/vfs/README.md

sqlite-wasm-rs stores data in memory by default, and you can also choose OPFS. https://github.com/Spxg/sqlite-wasm-rs/blob/master/sqlite-wasm-rs/tests/allin/vfs/sahpool.rs

Spxg commented:
Compared with the previous changes, I deleted `precompiled-wasm` and `modern-full-wasm32-unknown-unknown` features. Introduced `modern-base` features, which should be platform-independent.

I am not sure whether `rusqlite` has an MSRV rule, because the current MSRV of `sqlite-wasm-rs` is 1.78.0, which may break the rule of `rusqlite`. @gwenn 

The CI process has been run on my branch: <https://github.com/Spxg/rusqlite/actions/runs/16076570750>.
At the same time, `sqlite-wasm-rs` will run `rusqlite` tests for each commit to ensure that the function runs normally: <https://github.com/Spxg/sqlite-wasm-rs/actions/runs/16076725506/job/45373583850>

`sqlite-wasm-rs` has been running stably in `diesel`, `sqlight`, and my private project for the past six months. However, not all scenarios are suitable for using ORM or directly using SQLite C API to meet requirements, `rusqlite` can make up for these scenarios. I hope it can be merged.

gwenn commented:
> I am not sure whether `rusqlite` has an MSRV rule, because the current MSRV of `sqlite-wasm-rs` is 1.78.0, which may break the rule of `rusqlite`. @gwenn

You cannot break the `rusqlite` MSRV:
https://github.com/rusqlite/rusqlite?tab=readme-ov-file#minimum-supported-rust-version-msrv

Spxg commented:
> except by targeting a beta/nightly version

The `MSRV` of `sqlite-wasm-rs` is  version `1.78.0` stable currently, so there is no problem.
https://github.com/Spxg/sqlite-wasm-rs?tab=readme-ov-file#minimum-supported-rust-version-msrv

Spxg commented:
@gwenn any suggestions? `wasm-bindgen` was unmaintained for a while, which might have been a concern for you, but the organization has changed and things are moving in a positive direction. https://github.com/wasm-bindgen/wasm-bindgen/issues/4533

gwenn commented:
Just for reference:
https://blog.rust-lang.org/2025/04/04/c-abi-changes-for-wasm32-unknown-unknown/

Spxg commented:
> Just for reference: https://blog.rust-lang.org/2025/04/04/c-abi-changes-for-wasm32-unknown-unknown/

Yes, `wasm32-unknown-unknown` is now C-ABI aligned, which means linking to `emscripten` or `wasi` will not be a problem.

Spxg commented:
> Which means the official [sqlite3.wasm](https://sqlite.org/download.html#wasm) ?

This means we can compile `sqlite3` with `emcc` and link it into our rust `wasm32-unknown-unknown` program, just like `sqlite-wasm-rs` does.

However, this is only one step, more work is actually to implement a persistent VFS under `wasm32-unknown-unknown`, like opfs: https://github.com/Spxg/sqlite-wasm-rs/blob/master/sqlite-wasm-rs/src/vfs/sahpool.rs

 If you are interested, you can try it out at https://sqlight.dev

gwenn commented:
Ok, thanks for the details.
